<!-- RGB-D Visual odometry -->

<launch>

  # ORB, SURF, GTF, STAR
  <arg name="detector_type" default="GFT"/> 

  # ICPProbModel
  <arg name="reg_type" default="ICPProbModel"/> 

  <node pkg="ccny_rgbd" type="visual_odometry_node" name="visual_odometry_node" 
    output="screen">
    
    <!-- NOTE: if using data from OpenNI driver directly, (without 
    ccny_rgbd/rgbd_image_proc"), then add the following remappings. 
    Also add these to keyframe_mapper in vo+mapping.launch.

    <remap from="/rgbd/depth" to="/camera/depth_registered/image_rect_raw"/>
    <remap from="/rgbd/rgb"   to="/camera/rgb/image_rect_color"/>
    <remap from="/rgbd/info"  to="/camera/rgb/camera_info"/>
    
    -->
    
    #### diagnostics ##################################
        
    <param name="verbose"     value="true"/>    
    
    #### frames and tf output #########################
    
    <param name="publish_tf"  value="true"/>
    <param name="fixed_frame" value="/odom"/>
    <param name="base_frame"  value="/camera_link"/>
       
    #### features #####################################
    
    #  ORB, SURF, or GFT (Good features to track)
    <param name="feature/detector_type"               value="$(arg detector_type)"/> 
    <param name="feature/smooth"                      value="0"/>
    <param name="feature/max_range"                   value="7.0"/>
    <param name="feature/max_stdev"                   value="0.05"/>
    <param name="feature/show_keypoints"              value="false"/>
    <param name="feature/publish_feature_cloud"       value="true"/>
    <param name="feature/publish_feature_covariances" value="false"/>

    #### features: GFT ################################

    <param name="feature/GFT/n_features"   value = "400"/>
    <param name="feature/GFT/min_distance" value = "2.0"/>

    #### features: SURF ###############################
  
    <param name="feature/SURF/threshold" value = "400"/>

    #### features: ORB ###############################
  
    <param name="feature/ORB/n_features" value = "300"/>
    <param name="feature/ORB/threshold"  value = "31"/>

    #### registration #################################

    <param name="reg/reg_type"          value="$(arg reg_type)"/>
    <param name="reg/motion_constraint" value="2"/>

    #### registration: ICP Prob Model #################

    <param name="reg/ICPProbModel/max_iterations"            value="10"/>
    <param name="reg/ICPProbModel/max_model_size"            value="5000"/>
    <param name="reg/ICPProbModel/n_nearest_neighbors"       value="4"/>
    <param name="reg/ICPProbModel/max_assoc_dist_mah"        value="10.0"/>
    <param name="reg/ICPProbModel/max_corresp_dist_eucl"     value="0.15"/>
    <param name="reg/ICPProbModel/publish_model_cloud"       value="true"/>
    <param name="reg/ICPProbModel/publish_model_covariances" value="false"/>
  </node>

</launch>

